File: Runtime/python/Phrost/Live-Reload.py
import os
import pathlib
import sys
from typing import Any, Callable, Dict, NoReturn, Optional

from ChannelPacker import ChannelPacker
from Channels import Channels
from Events import Events

# This is the path used in poll() which is relative to this file
# (equivalent to __DIR__ . "/../save.data")
_HARDCODED_SAVE_PATH = pathlib.Path(__file__).parent / "../save.data"


class LiveReload:
    def __init__(
        self,
        shutdown_flag_path: str,
        save_path: str,
        # --- ADD THESE ARGUMENTS ---
        # We type-hint 'Callable' which is Python's way of saying 'function'
        wake_func: Callable[[bytes], None],
        sleep_func: Callable[[], bytes],
    ):
        self.shutdown_flag_path: Optional[str] = shutdown_flag_path
        self.save_path: Optional[str] = save_path
        self.reset_pending: bool = False

        # --- STORE THE FUNCTIONS ON SELF ---
        self.phrost_wake = wake_func
        self.phrost_sleep = sleep_func

    def set_paths(self, shutdown_flag_path: str, save_path: str):
        """
        Re-injects paths after potential un-pickling (unserialization).
        """
        self.shutdown_flag_path = shutdown_flag_path
        self.save_path = save_path

    def is_reset_pending(self) -> bool:
        return self.reset_pending

    def poll(self, is_loaded: bool = False):
        # This logic uses a hard-coded path, matching the PHP file
        if not is_loaded and os.path.isfile(_HARDCODED_SAVE_PATH):
            try:
                with open(_HARDCODED_SAVE_PATH, "rb") as f:
                    save_content = f.read()

                # Check if content was read successfully AND is not empty
                if save_content:
                    self.phrost_wake(save_content)
                else:
                    print("Save.data file was empty. Starting fresh.")
                    # Delete the bad file so we don't try again
                    os.remove(_HARDCODED_SAVE_PATH)

            except IOError as e:
                print(f"Save.data file was unreadable: {e}. Starting fresh.")
                try:
                    os.remove(_HARDCODED_SAVE_PATH)
                except OSError:
                    pass  # Failed to delete, but we'll proceed anyway

        if self.shutdown_flag_path and os.path.isfile(self.shutdown_flag_path):
            # Read the flag's content (e.g., "reset" or "save")
            try:
                with open(self.shutdown_flag_path, "r") as f:
                    flag_content = f.read().strip()
            except IOError:
                flag_content = ""  # Treat as "save" if unreadable

            # Delete the flag *immediately* so we don't loop
            try:
                os.remove(self.shutdown_flag_path)
            except OSError as e:
                print(f"Warning: Could not delete shutdown flag: {e}")

            if flag_content == "reset":
                print("Hard reset detected. Skipping save and unloading.")
                # We just exit to force a reload. The *next* load will be clean
                # because reset() already deleted save.data.
                print("unloading")  # <-- THIS IS THE FIX
                sys.exit(0)  # Use sys.exit(0) for a clean exit
            else:
                # Flag was empty, "save", or anything else:
                print("Saving state before unloading...")
                if self.save_path:
                    try:
                        save_data = self.phrost_sleep()
                        with open(self.save_path, "wb") as f:
                            f.write(save_data)
                    except IOError as e:
                        print(f"Error writing save file: {e}")

                # This is the magic string the engine needs to see
                print("unloading")
                sys.exit(0)

    def reset_on_event(self, event: Dict[str, int], keycode: int, mod: int):
        """
        Triggers on matching event, the reset of world data
        """
        if event.get("keycode") == keycode and (event.get("mod", 0) & mod):
            print("Hard Reset Triggered! Pending for next frame.")
            self.reset_pending = True

    def reset(self, world: Dict[str, Any], packer: ChannelPacker):
        # Unload sprites and bodies before reset.
        for sprite_id, sprite in world.get("sprites", {}).items():
            packer.add(
                Channels.RENDERER.value,
                Events.SPRITE_REMOVE,
                [sprite.id0, sprite.id1],
            )

        for physics_body_id, physics_body in world.get("physicsBodies", {}).items():
            packer.add(
                Channels.PHYSICS.value,
                Events.PHYSICS_REMOVE_BODY,
                [physics_body.id0, physics_body.id1],
            )

        # This method uses the configurable self.save_path
        if self.save_path and os.path.isfile(self.save_path):
            try:
                os.remove(self.save_path)
            except OSError as e:
                print(f"Could not remove save file: {e}")

        # Create the shutdown flag with "reset" content
        if self.shutdown_flag_path:
            try:
                with open(self.shutdown_flag_path, "w") as f:
                    f.write("reset")
            except IOError as e:
                print(f"Could not write reset flag: {e}")

        # We are no longer pending
        self.reset_pending = False

    def shutdown_on_event(self, event: Dict[str, int], keycode: int, mod: int):
        """
        Triggers on matching event, the shutdown of game engine
        """
        if event.get("keycode") == keycode and (event.get("mod", 0) & mod):
            print("Shutting down")
            self.shutdown()

    def shutdown(self) -> NoReturn:
        """
        Hard coded value to trigger shutdown from outside process
        """
        sys.exit(10)

File: Runtime/javascript/Phrost/LiveReload.js
const fs = require("fs");
const path = require("path");
const { Events, Channels } = require("./Events");
const Mod = require("./Mod");

/**
 * LiveReload Class
 * * Handles hot-reloading and state management.
 */
class LiveReload {
  /**
   * @param {string} shutdownFlagPath
   * @param {string} savePath
   */
  constructor(shutdownFlagPath, savePath) {
    this.shutdownFlagPath = shutdownFlagPath;
    this.savePath = savePath;
    this.resetPending = false;
  }

  /**
   * Re-injects paths after unserialization (if applicable in JS).
   * * @param {string} shutdownFlagPath
   * @param {string} savePath
   */
  setPaths(shutdownFlagPath, savePath) {
    this.shutdownFlagPath = shutdownFlagPath;
    this.savePath = savePath;
  }

  isResetPending() {
    return this.resetPending;
  }

  /**
   * Polls for live reload signals.
   * Should be called once per frame.
   * * @param {boolean} [isLoaded=false]
   */
  poll(isLoaded = false) {
    // Note: Synchronous file ops are used here to match PHP behavior
    // and simplicity in the game loop structure.

    if (!isLoaded && fs.existsSync(this.savePath)) {
      // In JS, we assume Phrost_Wake and Phrost_Sleep are global or imported functions
      // handling the state restoration.
      try {
        const saveContent = fs.readFileSync(this.savePath, "utf8");
        if (saveContent && saveContent.trim().length > 0) {
          if (typeof global.Phrost_Wake === "function") {
            global.Phrost_Wake(saveContent);
          }
        } else {
          console.log(
            "Save.data file was empty or unreadable. Starting fresh.",
          );
          fs.unlinkSync(this.savePath);
        }
      } catch (e) {
        console.error("Error reading save file:", e);
      }
    }

    if (fs.existsSync(this.shutdownFlagPath)) {
      try {
        const flagContent = fs
          .readFileSync(this.shutdownFlagPath, "utf8")
          .trim();

        // Delete flag immediately
        fs.unlinkSync(this.shutdownFlagPath);

        if (flagContent === "reset") {
          console.log("Hard reset detected. Skipping save and unloading.");
          // Unloading signal
          console.log("unloading");
          process.exit(0); // Exit code 10 for reload/shutdown
        } else {
          console.log("Saving state before unloading...");
          if (typeof global.Phrost_Sleep === "function") {
            const state = global.Phrost_Sleep();
            fs.writeFileSync(this.savePath, state);
          }

          console.log("unloading");
          process.exit(0);
        }
      } catch (e) {
        console.error("Error processing shutdown flag:", e);
      }
    }
  }

  /**
   * Triggers reset on event match.
   * * @param {Object} event
   * @param {number} keycode
   * @param {number} mod
   */
  resetOnEvent(event, keycode, mod) {
    if (event.keycode === keycode && event.mod & mod) {
      console.log("Hard Reset Triggered! Pending for next frame.");
      this.resetPending = true;
    }
  }

  /**
   * Performs a hard reset of the application state.
   * * @param {Object} world - World object containing sprites and bodies.
   * @param {Object} packer - ChannelPacker
   */
  reset(world, packer) {
    // Unload sprites
    if (world.sprites) {
      for (const spriteId in world.sprites) {
        const sprite = world.sprites[spriteId];
        // Check if sprite has remove method or construct event manually
        if (typeof sprite.remove === "function") {
          // Assuming sprite has remove method, otherwise:
          packer.add(Channels.RENDERER, Events.SPRITE_REMOVE, [
            sprite.id0,
            sprite.id1,
          ]);
        } else {
          packer.add(Channels.RENDERER, Events.SPRITE_REMOVE, [
            sprite.id0,
            sprite.id1,
          ]);
        }
      }
    }

    // Unload physics bodies
    if (world.physicsBodies) {
      for (const bodyId in world.physicsBodies) {
        const body = world.physicsBodies[bodyId];
        packer.add(Channels.PHYSICS, Events.PHYSICS_REMOVE_BODY, [
          body.id0,
          body.id1,
        ]);
      }
    }

    if (fs.existsSync(this.savePath)) {
      fs.unlinkSync(this.savePath);
    }

    fs.writeFileSync(this.shutdownFlagPath, "reset");
    this.resetPending = false;
  }

  /**
   * Triggers shutdown on event match.
   * * @param {Object} event
   * @param {number} keycode
   * @param {number} mod
   */
  shutdownOnEvent(event, keycode, mod) {
    let modMatch = false;
    modMatch = mod === Mod.NONE ? true : event.mod & mod;

    if (event.keycode === keycode && modMatch) {
      console.log("Shutting down");
      this.shutdown();
    }
  }

  /**
   * Triggers immediate hard shutdown.
   */
  shutdown() {
    process.exit(10);
  }
}

module.exports = LiveReload;
